### DISCUSSION TITLE
CCSD with custom Hamiltonian and mo_energies

### DISCUSSION THREAD

**Discussion by Physfock** [pyscf/pyscf#2650] [Q&A] [answered]
As far as I understand, inside CCSD there is always a recalculation of the SCF and the energies of the molecular orbitals, which are necessary to perform CCSD cycles (denominator calculations). I would like to bypass the SCF calculations inside CCSD again by simply sending the already known result to the input

mf = scf.RHF(mol)
mf.mo_coeff = C_new
mf.mo_energy = E_new
mf.e_tot = E_tot_new  --> let say  C_new,  E_new and E_tot_new are calculated by another program

 --> Below I define custom hamiltonian:

mf.get_hcore = lambda *args: F_new
mf.get_ovlp  = lambda *args: S_new
mf._eri      = ao2mo.restore(1, eri_new, mol.nao)  

 --> If I now do mf.kernel() I get new C'_new, E'_new and  E'_tot_new which differs from C_new, E_new and  E_tot_new but I want to avoid this.

 --> Perform the CCSD calculation using the new mean-field object

mycc = cc.RCCSD(mf, frozen = None, mo_coeff = C_new, mo_occ = mf.mo_occ) --> just in case, I’m once again sending new coefficients to the arguments, but not new mo_energies E_new since there is no corresponding attributes (and it seems that CCSD just calculates them inside starting new SCF cycle which as I said above gives incorrect E'_new for my particular case)

mycc.kernel()

In other words, I just want to run RCCSD with predefined parameters, without any recalculation inside. Maybe it's possible to somehow assign new values to variables globally?

**Comment by MatthewRHermes** [comment, marked as answer]
It doesn't do an entire SCF iteration; just one Fock build and sets `mo_energy` to the diagonal elements. I don't think it changes the MO coefficients or the SCF total energy. I think the member function you want to interdict is `ao2mo`: 
```
class myCCSD (mycc.__class__):
    def ao2mo (self, mo_coeff=None):
        eris = super ().ao2mo (mo_coeff=mo_coeff)
        eris.mo_energy = my_mo_energy
        return eris
from pyscf import lib
my_patched_cc = lib.view (mycc, myCCSD)
my_patched_cc.kernel ()
```
